### DISCUSSION TITLE
Error initializing I2C

### DISCUSSION THREAD

**Discussion by grender007** [pschatzmann/arduino-audio-driver#5] [Q&A] [answered]
Hi, I've been working on my code for a couple of weeks (using ES8388 as an ADC for ESP32 S3). Most of the time, I didn't get anything from StreamCopy and I kind of figured why. There is some error in the I2C communication, but I'm not exactly sure what it could be.

Info:    output_device: 0
Error:   ->p_wire->endTransmission: 4
Error:   ->p_wire->endTransmission: 4
Error:   ->p_wire->endTransmission: 4
Error:   ->p_wire->endTransmission: 4
Error:   ->p_wire->endTransmission: 4
Error:   ->p_wire->endTransmission: 4
Error:   ->p_wire->endTransmission: 4
Error:   ->p_wire->endTransmission: 4
Error:   ->p_wire->endTransmission: 4
Error:   ->p_wire->endTransmission: 4
Error:   init failed

I attached my code below:

[test.txt](https://github.com/pschatzmann/arduino-audio-driver/files/14336300/test.txt)

**Comment by pschatzmann** [comment, marked as answer]
Since the ES8388 is working perfectly with the AudioKit, it can only be the initialization of I2C. Are you doing this via the pin definitions or yourself ? Both options are explained in the Wiki

**Reply by grender007** [reply]
I'm using the pin definitions. (This is the line)

 my_pins.addI2C(audio_driver::PinFunction::CODEC, 4, 3, 0x10, 100000, Wire);

**Comment by pschatzmann** [comment]
Did you try with some other pins ? The default (recommended?) pins would be 
- GPIO 8 (SDA)
- GPIO 9 (SCL)

**Reply by grender007** [reply]
I just tried to change the pins, but I got the same error:

Error:   ->p_wire->endTransmission: 4
Error:   ->p_wire->endTransmission: 4
Error:   ->p_wire->endTransmission: 4
Error:   ->p_wire->endTransmission: 4
Error:   ->p_wire->endTransmission: 4
Error:   ->p_wire->endTransmission: 4
Error:   ->p_wire->endTransmission: 4
Error:   ->p_wire->endTransmission: 4
Error:   ->p_wire->endTransmission: 4
Error:   ->p_wire->endTransmission: 4
Info:    output_device: 3
Error:   ->p_wire->endTransmission: 4
Error:   ->p_wire->endTransmission: 4
Error:   ->p_wire->endTransmission: 4
Error:   ->p_wire->endTransmission: 4
Error:   ->p_wire->endTransmission: 4
Error:   ->p_wire->endTransmission: 4
Error:   ->p_wire->endTransmission: 4
Error:   ->p_wire->endTransmission: 4
Error:   ->p_wire->endTransmission: 4
Error:   ->p_wire->endTransmission: 4
Error:   init failed

I also used one of the default WireScan codes to verify the I2C connection and it apparently worked. Now, I'm trying with the default pins

**Reply by pschatzmann** [reply]
You can try to set the log mode to debug to see in detail what's going on.
You can also run the i2c port scan sketch: What is the address that you get ?
Please note that the 0x10 is just some information which is not used. The port is hardcoded with
#define ES8388_ADDR 0x20  
This would correspond with 0x10 in Arduino naming.

**Reply by grender007** [reply]
I used the scan sketch to verify the I2C connection, and I showed that the codec had the 0x10 address. Should I change it to 0x20 anyways?
Edit 1: I just did it (0x20 address with default pins for I2C) and still got the same error
Edit 2: I got this from the scan

Scanning for I2C devices ...
I2C device found at address 0x10

**Reply by pschatzmann** [reply]
Set the log [level to debug](https://github.com/pschatzmann/arduino-audio-driver/wiki/Logging) and you can also try to reduce the I2C speed.
I have no idea what else could be wrong....

**Reply by pschatzmann** [reply]
With the latest commits, I changed the I2C address logic to use Arduino Conventions: This should be less confusing, since it the log you will see the values that you expect.

**Reply by pschatzmann** [reply]
endTransmission returns: byte , which indicates the status of the transmission:
0: success.
1: busy timeout upon entering endTransmission()
2: START bit generation timeout.
3: end of address transmission timeout.
4: data byte transfer timeout.
5: data byte transfer succeeded, busy timeout immediately after.
